new/cookbook-segment-deprecation.md

```diff
@@ -0,0 +1,51 @@
+---
+RFC: unassigned
+Author: Lamont Granquist <lamont@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Title
+
+Cookbook Segment Deprecation
+
+## Motivation
+
+    As a Chef User/Developer,
+    I want to be able to extend the structure of cookbooks without fighting with a wire protocol
+    written 6 years ago,
+    Because nobody wants that.
+
+## Specification
+
+This RFC will add a special `all_files` segment to cookbooks which will duplicate (initially)
+all of the existing cookbook manifest along with arbitrary files outside of the current scope of
+any of the cookbook segments.  The Chef Server will need to be patched in order to support this
+new 'segment' and accept it since the Chef Server currently does strict enforcement of file
+segments server-side.  This will also generate a version bump of the Chef Server protocol which
+will be used by clients to determine if the Chef Server supports this field or not.
+
+New knife clients that support this segment will use the Chef Server protocol version to determine
+if they MUST upload cookbooks using the old segment format, or if they SHOULD upload using only
+the `all_files` format.
+
+For backwards compatibility, the Chef Server MUST serve cookbooks with both standard segments and the
+new `all_files` segment unless it can determine that the client prefers one or other.  If it can
```

Comment:
I believe we should be able to use the API version here as well.  For versions before this was introduced or no-version we return the current response.  For versions after it was introduce, serve the new response.  That is, I'm unclear on the cases where we would be unable to determine client preference.

What should the server do with all_files members that don't map to the old segments?